fuzz-tests: add test for amount-{sat, msat} arithmetic

Chand-ra opened the pull request:
The fuzz-amount test doesn't fuzz the arithmetic operations for `struct amount_sat` and `struct amount_msat`. Add a test for the same.

## Checklist
Before submitting the PR, ensure the following tasks are completed. If an item is not applicable to your PR, please mark it as checked:

- [x] The changelog has been updated in the relevant commit(s) according to the [guidelines](https://docs.corelightning.org/docs/coding-style-guidelines#changelog-entries-in-commit-messages).
- [x] Tests have been added or modified to reflect the changes.
- [x] Documentation has been reviewed and updated as needed.
- [x] Related issues have been listed and linked, including any that this PR closes.

Commit:
fuzz-tests: add test for amount-{sat, msat} arithmetic

Changelog-None: The `fuzz-amount` test doesn't fuzz the arithmetic
operations for `struct amount_sat` and `struct amount_msat`. Add a
test for them.

morehouse commented:
> Has the fuzzer found an actual bug or is it just another impossible edge case?

I think technically these are bugs -- `amount_msat_scale` and `amount_sat_scale` are supposed to return `false` when scaling fails and already do this for the positive-overflow case.  Since there's no usage comment requiring `scale` to be positive, probably the functions should handle negative overflow as well.

In practice, all current usages of these functions in the codebase have positive values for `scale`, so the bug can never manifest in the existing codebase.

morehouse commented:
> Makes me wonder why `scale`'s type isn't unsigned in the first place...

Probably because there's no way to represent an unsigned float.

Chand-ra commented:
Hey @morehouse,

As we discussed over email, I spent some time last week investigating the potential bug caused by an assertion failure in `amount_msat_sub_fee()` to determine whether it could be triggered externally. I grepped the codebase for all callers of `amount_msat_sub_fee()` and found that the only caller resides in `plugins/askrene/refine.c`. This, in turn, ultimately traces back to askrene’s `getroutes` command.

To test this path, I added a user-level Python test in `tests/test_askrene.py` to try passing the error-causing parameters to `amount_msat_sub_fee()`. I found that any parameter values that get too large are truncated to a set of ‘safe values’, making it impossible to reproduce the breakage through this command.

In summary, it doesn’t appear that any external message can trigger the failure seen in the fuzz test. I’ve pushed the test in case you’d like to take a look.

Commit:
fuzz-tests: Add a seed corpus for the new test

Add a minimal input set as a seed corpus for the newly introduced
test. This leads to discovery of interesting code paths faster.

Chand-ra commented:
The test works without any breakage now and is ready to be merged.

morehouse reviewed (commented):
I am hesitant to merge this fuzz target, since it basically has to reimplement all the operations in `common/amount.c` so we have something to compare against.  It seems a bit silly when we could just directly review the implementations in `common/amount.c`.

That said, the target did find a [couple cases](https://github.com/ElementsProject/lightning/pull/8306) of undefined behavior, which makes me think it would be worth merging this target at least as regression tests for those bugs.

Chand-ra commented on tests/fuzz/fuzz-amount-arith.c:
```diff
@@ -0,0 +1,295 @@
+static struct amount_sat fromwire_amount_sat_bounded(const u8 **cursor, size_t *max)
+{
+	struct amount_sat amt = fromwire_amount_sat(cursor, max);
+	 amt.satoshis %= (MAX_BTC + 1);
```
Right, but isn't it a valid assumption that in production we won't be doing arithmetic on anything greater than the greatest possible value for satoshis/millisatoshis?

Chand-ra commented on tests/fuzz/fuzz-amount-arith.c:
```diff
@@ -0,0 +1,295 @@
+        if (amount_msat_fee(&fee, output, fee_base, fee_prop)) {
+            struct amount_msat sum;
+            if (amount_msat_add(&sum, output, fee))
+                assert(amount_msat_less_eq(sum, input));
```
In `amount_msat_fee`'s [defintion](https://github.com/ElementsProject/lightning/blob/2e7181d04fe29e731a10cde0f5d600c666d8eb68/common/amount.c#L578):
```
/* BOLT #7:
*
*   - SHOULD accept HTLCs that pay a fee equal to or greater than:
*    - fee_base_msat + ( amount_to_forward * fee_proportional_millionths / 1000000 )
*/
```
So it is valid for `sum` to be <= `input`.

Chand-ra commented:
The corpus for this test is now generated with the fix in #8306 applied, so that PR needs to be merged before this one can.

madelinevibes commented:
@Chand-ra can you resolve the conflicts and clear CI by the end of this week so we can add this to 26.06? Release candidate planned for Monday 11 May. 

If you can't we'll move it to 25.09

Chand-ra commented:
> @Chand-ra can you resolve the conflicts and clear CI by the end of this week so we can add this to 26.06? Release candidate planned for Monday 11 May.
> 
> If you can't we'll move it to 25.09

I'll try to, and let you know of any updates by the end of the week.

Commit:
Makefile: allow raw amount access in fuzz tests

Updates the `check-amount-access` target to exclude `tests/fuzz/*`
so fuzzing tests can directly verify .satoshis and .millisatoshis
without failing CI.

Chand-ra commented:
Hey @madelinevibes, all done here!

sangbida reviewed (approved):
Looks good :)
